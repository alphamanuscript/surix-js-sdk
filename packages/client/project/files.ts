import { AxiosInstance } from 'axios';
import { ApiFile, ProjectFiles } from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<ApiFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return res.data;
    },
    async list (): Promise<ApiFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return res.data;
    }
  };
}