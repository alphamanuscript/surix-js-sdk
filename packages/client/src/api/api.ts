import axios, { AxiosInstance } from 'axios';

export function getApiClient (baseUrl: string, apiKey: string): AxiosInstance {
  const axiosInstance = axios.create({
    baseURL: baseUrl
  })
  axiosInstance.defaults.headers['Authorization'] = `apikey ${apiKey}`
  return axiosInstance
}