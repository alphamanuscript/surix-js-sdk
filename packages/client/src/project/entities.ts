import { 
  ApiEntity,  
  expandEntity, 
  expandQuery, 
  wrapEntity, 
  wrapEntityArray, 
  WrappedEntity
} from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import {  DeletedEntities, Entity, ProjectEntities, Query } from '../types';

export function getProjectEntities (projectId: string, apiClient: AxiosInstance): ProjectEntities {
  return {
    async get(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return wrapEntity(res.data);
    },
    async query(query: Query = {}): Promise<WrappedEntity[]> {
      const expandedQuery = expandQuery(query);
      if (Object.keys(expandedQuery.query).length === 0) {
        delete expandedQuery.query;
      }
      const res = await apiClient.post<ApiEntity[]>(
        `/projects/${projectId}/entities/query`, expandedQuery);
      const entities = res.data;
      return wrapEntityArray(entities);
    },
    async update(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity };
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.patch<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async put(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity }; // Make a shallow copy
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.put<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async delete(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.delete(`/projects/${projectId}/entities/${entityId}`);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async deleteMany(entityIds: string[]): Promise<DeletedEntities> {
      const res = await apiClient.delete(`/projects/${projectId}/entities`, 
      {
        data: {
          entities: entityIds
        }
      });
      return res.data;
    },
    async create(entity: Entity): Promise<WrappedEntity> {
      const expandedEntity = expandEntity(entity);
      const res = await apiClient.post<ApiEntity>(
        `/projects/${projectId}/entities`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    }
  };
}