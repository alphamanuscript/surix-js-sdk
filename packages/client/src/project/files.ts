import {
  ApiFile,
  wrapFile,
  wrapFileArray,
  WrappedFile } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectFiles} from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<WrappedFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return wrapFile(res.data);
    },
    async list (): Promise<WrappedFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return wrapFileArray(res.data);
    }
  };
}