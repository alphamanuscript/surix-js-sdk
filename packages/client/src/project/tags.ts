import { AxiosInstance } from 'axios';
import { ProjectTags, Tag, TagList  } from '../types';

export function getProjectTags (projectId: string, apiClient: AxiosInstance): ProjectTags {
  return {
    async list (): Promise<TagList> {
      const res = await apiClient.get<TagList>(`/projects/${projectId}/tags`);
      return res.data;
    }, 
    async rename (oldName: string, newName: string): Promise<Tag> {
      const res = await apiClient.put<Tag>(`/projects/${projectId}/tags/${oldName}`, {
        name: newName
      });
      return res.data;
    }
  };
}