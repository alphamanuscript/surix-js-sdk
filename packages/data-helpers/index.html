<script src="./dist/data-helpers.min.js"></script>