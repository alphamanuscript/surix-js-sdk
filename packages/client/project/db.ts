import { ApiEntity, EntityWrapper } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectDatabase, Query, TagList } from '../types';

export function getProjectDb (projectId: string, apiClient: AxiosInstance): ProjectDatabase {
  return {
    async get (entityId: string): Promise<EntityWrapper> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return new EntityWrapper(res.data);
    },
    async query(query: Query): Promise<EntityWrapper[]> {
      const res = await apiClient.post<ApiEntity[]>(`/projects/${projectId}/entities/query`, query);
      const entities = res.data;
      return entities.map(entity => new EntityWrapper(entity));
    },
    async getTags (): Promise<TagList[]> {
      const res = await apiClient.get<TagList[]>(`/projects/${projectId}/tags`);
      return res.data;
    }
  };
}