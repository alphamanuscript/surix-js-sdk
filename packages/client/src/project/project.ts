import { AxiosInstance } from 'axios';
import { getProjectDb } from './db';
import { getProjectFiles } from './files';

import { Project } from '../types';

export function getProjectApi (projectId: string, apiClient: AxiosInstance): Project {
  return {
    db: getProjectDb(projectId, apiClient),
    files: getProjectFiles(projectId, apiClient)
  };
}