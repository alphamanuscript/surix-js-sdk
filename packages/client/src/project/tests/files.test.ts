import * as dataHelpers from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import * as api from '../../api';
import { getProjectApi } from '../project';

describe('Project Files', () => {
  let apiFiles: dataHelpers.ApiFile[];
  beforeEach(() => {
    apiFiles = [
      {
        _id: 'f1',
        name: 'F1',
        filename: 'f1.jpeg',
        mimeType: 'image/jpeg',
        status: 'ready',
        size: 200,
        public: true,
        downloadUrl: 'http://download/f1',
        uploadUrl: 'http://upload/f1',
        tags: ['t1'],
        createdBy: { _id: 'u1', type: 'user' },
        createdAt: '2017-02-22T09:19:33.885Z',
        updatedAt: '2018-10-22T09:19:33.885Z',
      },
      {
        _id: 'f2',
        name: 'F2',
        filename: 'f2.jpeg',
        mimeType: 'image/jpeg',
        status: 'pending',
        size: 0,
        public: true,
        downloadUrl: '',
        uploadUrl: 'http://upload/f1',
        tags: ['t1'],
        createdBy: { _id: 'u1', type: 'user' },
        createdAt: '2017-02-22T09:19:33.885Z',
        updatedAt: '2018-10-22T09:19:33.885Z',
      },
    ];
  });

  describe('get', () => {
    let apiClient: AxiosInstance;
    function callMockedGet (): Promise<dataHelpers.WrappedFile> {
      apiClient = api.getApiClient('http://baseurl');
      jest.spyOn(apiClient, 'get').mockReturnValue(
        Promise.resolve({ data: apiFiles[0] })
      );
      jest.spyOn(dataHelpers, 'wrapFile');
      const project = getProjectApi('project1', apiClient);
      return project.files.get('f1');
    }
    it('should call the GET /project/:pid/files/:id endpoint', async () => {
      await callMockedGet();
      expect(apiClient.get).toHaveBeenCalledWith(`/projects/project1/files/f1`);
    });
    it('should return the wrapped file', async () => {
      const file = await callMockedGet();
      expect(dataHelpers.wrapFile).toHaveBeenCalledWith(apiFiles[0]);
      expect(file).toMatchSnapshot();
    });
  });

  describe('list', () => {
    let apiClient: AxiosInstance;
    function callMockedList (): Promise<dataHelpers.WrappedFile[]> {
      apiClient = api.getApiClient('http://baseurl');
      jest.spyOn(apiClient, 'get').mockReturnValue(
        Promise.resolve({ data: apiFiles })
      );
      jest.spyOn(dataHelpers, 'wrapFileArray');
      const project = getProjectApi('project1', apiClient);
      return project.files.list();
    }
    it('should call the GET /project/:pid/files endpoint', async () => {
      await callMockedList();
      expect(apiClient.get).toHaveBeenCalledWith(`/projects/project1/files`);
    });
    it('should return the wrapped file array', async () => {
      const files = await callMockedList();
      expect(dataHelpers.wrapFileArray).toHaveBeenCalledWith(apiFiles);
      expect(files).toMatchSnapshot();
    });
  });
});