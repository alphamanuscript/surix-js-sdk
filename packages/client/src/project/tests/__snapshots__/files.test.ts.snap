// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Files get should return the wrapped file 1`] = `
Object {
  "_id": "f1",
  "createdAt": 2017-02-22T09:19:33.885Z,
  "createdBy": Object {
    "_id": "u1",
    "type": "user",
  },
  "downloadUrl": "http://download/f1",
  "filename": "f1.jpeg",
  "id": "f1",
  "mimeType": "image/jpeg",
  "name": "F1",
  "public": true,
  "rawFile": Object {
    "_id": "f1",
    "createdAt": "2017-02-22T09:19:33.885Z",
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "http://download/f1",
    "filename": "f1.jpeg",
    "mimeType": "image/jpeg",
    "name": "F1",
    "public": true,
    "size": 200,
    "status": "ready",
    "tags": Array [
      "t1",
    ],
    "updatedAt": "2018-10-22T09:19:33.885Z",
    "uploadUrl": "http://upload/f1",
  },
  "size": 200,
  "status": "ready",
  "tags": Array [
    "t1",
  ],
  "updatedAt": 2018-10-22T09:19:33.885Z,
  "uploadUrl": "http://upload/f1",
}
`;

exports[`Project Files list should return the wrapped file 1`] = `
Array [
  Object {
    "_id": "f1",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "http://download/f1",
    "filename": "f1.jpeg",
    "id": "f1",
    "mimeType": "image/jpeg",
    "name": "F1",
    "public": true,
    "rawFile": Object {
      "_id": "f1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "downloadUrl": "http://download/f1",
      "filename": "f1.jpeg",
      "mimeType": "image/jpeg",
      "name": "F1",
      "public": true,
      "size": 200,
      "status": "ready",
      "tags": Array [
        "t1",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f1",
    },
    "size": 200,
    "status": "ready",
    "tags": Array [
      "t1",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f1",
  },
  Object {
    "_id": "f2",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "",
    "filename": "f2.jpeg",
    "id": "f2",
    "mimeType": "image/jpeg",
    "name": "F2",
    "public": true,
    "rawFile": Object {
      "_id": "f2",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "downloadUrl": "",
      "filename": "f2.jpeg",
      "mimeType": "image/jpeg",
      "name": "F2",
      "public": true,
      "size": 0,
      "status": "pending",
      "tags": Array [
        "t1",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f1",
    },
    "size": 0,
    "status": "pending",
    "tags": Array [
      "t1",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f1",
  },
]
`;

exports[`Project Files list should return the wrapped file array 1`] = `
Array [
  Object {
    "_id": "f1",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "http://download/f1",
    "filename": "f1.jpeg",
    "id": "f1",
    "mimeType": "image/jpeg",
    "name": "F1",
    "public": true,
    "rawFile": Object {
      "_id": "f1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "downloadUrl": "http://download/f1",
      "filename": "f1.jpeg",
      "mimeType": "image/jpeg",
      "name": "F1",
      "public": true,
      "size": 200,
      "status": "ready",
      "tags": Array [
        "t1",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f1",
    },
    "size": 200,
    "status": "ready",
    "tags": Array [
      "t1",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f1",
  },
  Object {
    "_id": "f2",
    "createdAt": 2017-02-22T09:19:33.885Z,
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "downloadUrl": "",
    "filename": "f2.jpeg",
    "id": "f2",
    "mimeType": "image/jpeg",
    "name": "F2",
    "public": true,
    "rawFile": Object {
      "_id": "f2",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "downloadUrl": "",
      "filename": "f2.jpeg",
      "mimeType": "image/jpeg",
      "name": "F2",
      "public": true,
      "size": 0,
      "status": "pending",
      "tags": Array [
        "t1",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
      "uploadUrl": "http://upload/f1",
    },
    "size": 0,
    "status": "pending",
    "tags": Array [
      "t1",
    ],
    "updatedAt": 2018-10-22T09:19:33.885Z,
    "uploadUrl": "http://upload/f1",
  },
]
`;
